cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -94,6 +97,64 @@ Since policy revision numbers from Cilium are reset when the local Cilium instan
+
+We can run a gRPC server in the agent to serve the DNS rules to the DNS proxy. SDP will be responsible for creating the connection with the agent. And once SDP establish a connection agent can keep track of the stream and send the DNS rules to the SDP. Agent can then reuse the same stream to send updates in the DNS rules.
+In case, cilium agent is still not up, SDP will keep trying to connect to the agent until the connection is established. 
+
+##### Pros
+
+* SDP instances has the responsibility to connect to the agent.
+* Reusing the same stream will be effiecient in terms of resources as we are not creating/destroying the stream for every update.
+
+##### Cons
+
+* An overhead on the cilium agent to keep track of the streams of the connected SDP instances.
+* Streams are not thread safe, so if we have multiple threads using the same stream we will need to handle the synchronization.
+* If SDP in future decides to handle policy updates as well, it will be tricky. SDP needs to keep trying to sync with CA as well as CA needs to keep a track of the SDP instances to send updates.
+
+![standalone-dns-proxy-up-scenario](./images/standalone-dns-proxy-up-scenario.png)
+![standalone-dns-proxy-down-scenario](./images/standalone-dns-proxy-down-scenario.png)
+
+
+#### Option 2: Running the GRPC server in the SDP as well
+
+SDP can also runs a grpc server so that the agent can send the DNS rules to the SDP when the rules are ready instead of keeping alive the stream. On startup/restart SDP follows the same pattern as that of single grpc server, it asks for DNS rules from Cilium agent. For policy updates, agent just sends the data to server running at SDP.
+
+##### Pros
+
+* No overhead on the agent to keep a track of the open streams with the SDP.
+* Better fault isolation(one stream failing does not affect the other streams)
+
+##### Cons
+
+* Resource intensive as we are creating a new stream for every update.
+* If multiple SDP instance are running, we need to handle multiple grpc server on same address:port and send updates to each server from a agent.
+
+#### Option 3: Reading the ep_config.json file in the SDP
+
+We can read the DNS rules from the `ep_config.json` file. SDP will listen on the changes to the file and update the DNS rules accordingly. Cilium agent is the only one that writes to the file. Currently DNS rules are stored in the ep_config.json lazily as we only need those when DNS proxy restarts. We need to make sure that the DNS rules are updated in the file as soon as the policy is applied.
+Storing the dns rules will require identities associated with the selector labels as current dns proxy implementation uses that to filter the DNS traffic OR will need to change the way dns porxy filters the DNS requests.
+
+##### Pros
+
+* No need to run a gRPC server in the agent/SDP.
+* Removes the dependency on the agent to get the DNS rules.
+
+##### Cons
+
+* Will change the current implmentation/flow of flitering the dns requests.
```

Comment:
What changes wrt DNS requests ?

Reply:
I mean right now we filter based on the identities associated with the dns ip right ? If are writing to the file system similar to what is written currently i.e dns pod ip, the check would be to filter based on the ip rather than identities.
It depends on what we store in the file.

Reply:
Alright, after today's sync, we can consider updating this section to include relevant details from `ep_policy.json`